File: python/journal/lock.py
import os
import time


LOCK_VERSION = 'systemd-journal-sdk-lock-v1'
STALE_GRACE_SECONDS = 2.0


class WriterLock:
    def __init__(self, path, owner):
        self.path = path
        self.owner = owner

    @staticmethod
    def acquire(journal_path):
        lock_path = journal_path + '.lock'
        owner = _current_owner()
        while True:
            parent = os.path.dirname(lock_path)
            if parent:
                os.makedirs(parent, mode=0o750, exist_ok=True)
            try:
                fd = os.open(lock_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
            except FileExistsError:
                stale, holder = _lock_file_is_stale(lock_path)
                if not stale:
                    raise BlockingIOError(f'journal writer lock held by {holder}')
                try:
                    os.unlink(lock_path)
                except FileNotFoundError:
                    pass
                continue
            try:
                _write_owner(fd, owner)
                return WriterLock(lock_path, owner)
            except Exception:
                try:
                    os.close(fd)
                except OSError:
                    pass
                finally:
                    try:
                        os.unlink(lock_path)
                    except FileNotFoundError:
                        pass
                raise

    def release(self):
        if not self.path:
            return
        try:
            owner = _read_owner(self.path)
        except FileNotFoundError:
            self.path = None
            return
        if owner == _current_owner():
            try:
                os.unlink(self.path)
            except FileNotFoundError:
                pass
        self.path = None


def _write_owner(fd, owner):
    text = (
        f'{LOCK_VERSION}\n'
        f'pid={owner["pid"]}\n'
        f'boot_id={owner["boot_id"]}\n'
        f'start_time={owner["start_time"]}\n'
    ).encode('utf-8')
    os.write(fd, text)
    os.fsync(fd)
    os.close(fd)


def _lock_file_is_stale(path):
    try:
        owner = _read_owner(path)
    except Exception:
        try:
            age = time.time() - os.stat(path).st_mtime
        except FileNotFoundError:
            return True, 'missing lock'
        if age <= STALE_GRACE_SECONDS:
            return False, 'partially-created lock'
        return True, 'malformed stale lock'

    if owner.get('boot_id') != _boot_id():
        return True, f'pid {owner.get("pid")} from previous boot'
    try:
        start_time = _process_start_time(owner['pid'])
    except OSError:
        return True, f'stale pid {owner.get("pid")}'
    if start_time != owner.get('start_time'):
        return True, f'stale pid {owner.get("pid")}'
    return False, f'pid {owner.get("pid")}'


def _current_owner():
    pid = os.getpid()
    return {
        'pid': pid,
        'boot_id': _boot_id(),
        'start_time': _process_start_time(pid),
    }


def _boot_id():
    try:
        with open('/proc/sys/kernel/random/boot_id', 'r', encoding='ascii') as f:
            return f.read().strip()
    except OSError:
        return ''


def _process_start_time(pid):
    with open(f'/proc/{pid}/stat', 'r', encoding='ascii') as f:
        text = f.read()
    end = text.rfind(')')
    if end < 0:
        raise OSError(f'cannot parse /proc/{pid}/stat')
    fields = text[end + 2:].split()
    if len(fields) < 20:
        raise OSError(f'cannot parse start time from /proc/{pid}/stat')
    return fields[19]


def _read_owner(path):
    with open(path, 'r', encoding='utf-8') as f:
        lines = [line.strip() for line in f.readlines()]
    if len(lines) < 4 or lines[0] != LOCK_VERSION:
        raise ValueError('invalid lock metadata')
    owner = {}
    for line in lines[1:]:
        if '=' not in line:
            continue
        key, value = line.split('=', 1)
        if key == 'pid':
            owner[key] = int(value)
        elif key in ('boot_id', 'start_time'):
            owner[key] = value
    if owner.get('pid', 0) <= 0 or not owner.get('start_time'):
        raise ValueError('incomplete lock metadata')
    return owner

File: go/journal/lock.go
package journal

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	lockVersion       = "systemd-journal-sdk-lock-v1"
	lockStaleGrace    = 2 * time.Second
	lockFilePerm      = 0o600
	lockDirectoryPerm = 0o750
)

type writerLock struct {
	path  string
	owner lockOwner
}

// WriterLock is an optional cooperating-writer lock helper.
//
// The journal file format does not define a lock protocol. Callers that want
// SDK-to-SDK writer exclusion can acquire this helper before opening a Writer
// and release it after closing that Writer.
type WriterLock struct {
	lock *writerLock
}

// AcquireWriterLock acquires the optional cooperating-writer lock for path.
func AcquireWriterLock(journalPath string) (*WriterLock, error) {
	lock, err := acquireWriterLock(journalPath)
	if err != nil {
		return nil, err
	}
	return &WriterLock{lock: lock}, nil
}

// Release releases the optional cooperating-writer lock.
func (l *WriterLock) Release() error {
	if l == nil || l.lock == nil {
		return nil
	}
	err := l.lock.release()
	l.lock = nil
	return err
}

type lockOwner struct {
	PID       int
	BootID    string
	StartTime string
}

func acquireWriterLock(journalPath string) (*writerLock, error) {
	lockPath := journalPath + ".lock"
	owner, err := currentLockOwner()
	if err != nil {
		return nil, err
	}

	for {
		if err := os.MkdirAll(filepath.Dir(lockPath), lockDirectoryPerm); err != nil {
			return nil, err
		}
		f, err := os.OpenFile(lockPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, lockFilePerm)
		if err == nil {
			lock := &writerLock{path: lockPath, owner: owner}
			if err := writeLockOwner(f, owner); err != nil {
				_ = f.Close()
				_ = os.Remove(lockPath)
				return nil, err
			}
			if err := f.Close(); err != nil {
				_ = os.Remove(lockPath)
				return nil, err
			}
			return lock, nil
		}
		if !errors.Is(err, os.ErrExist) {
			return nil, err
		}

		stale, holder := lockFileIsStale(lockPath)
		if !stale {
			return nil, fmt.Errorf("journal writer lock held by %s", holder)
		}
		if err := os.Remove(lockPath); err != nil && !errors.Is(err, os.ErrNotExist) {
			return nil, err
		}
	}
}

func writeLockOwner(f *os.File, owner lockOwner) error {
	text := fmt.Sprintf("%s\npid=%d\nboot_id=%s\nstart_time=%s\n", lockVersion, owner.PID, owner.BootID, owner.StartTime)
	if _, err := f.WriteString(text); err != nil {
		return err
	}
	return f.Sync()
}

func (l *writerLock) release() error {
	if l == nil || l.path == "" {
		return nil
	}
	current, err := currentLockOwner()
	if err != nil {
		return err
	}
	owner, err := readLockOwner(l.path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			l.path = ""
			return nil
		}
		return err
	}
	if owner == current {
		err = os.Remove(l.path)
		if err != nil && !errors.Is(err, os.ErrNotExist) {
			return err
		}
	}
	l.path = ""
	return nil
}

func lockFileIsStale(path string) (bool, string) {
	owner, err := readLockOwner(path)
	if err != nil {
		info, statErr := os.Stat(path)
		if statErr == nil && time.Since(info.ModTime()) <= lockStaleGrace {
			return false, "partially-created lock"
		}
		return true, "malformed stale lock"
	}
	if owner.BootID != currentBootID() {
		return true, fmt.Sprintf("pid %d from previous boot", owner.PID)
	}
	start, err := processStartTime(owner.PID)
	if err != nil || start != owner.StartTime {
		return true, fmt.Sprintf("stale pid %d", owner.PID)
	}
	return false, fmt.Sprintf("pid %d", owner.PID)
}

func currentLockOwner() (lockOwner, error) {
	pid := os.Getpid()
	start, err := processStartTime(pid)
	if err != nil {
		return lockOwner{}, err
	}
	return lockOwner{PID: pid, BootID: currentBootID(), StartTime: start}, nil
}

func readLockOwner(path string) (lockOwner, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return lockOwner{}, err
	}
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	if len(lines) < 4 || lines[0] != lockVersion {
		return lockOwner{}, fmt.Errorf("invalid lock metadata")
	}
	var owner lockOwner
	for _, line := range lines[1:] {
		key, value, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		switch key {
		case "pid":
			pid, err := strconv.Atoi(value)
			if err != nil {
				return lockOwner{}, err
			}
			owner.PID = pid
		case "boot_id":
			owner.BootID = value
		case "start_time":
			owner.StartTime = value
		}
	}
	if owner.PID <= 0 || owner.StartTime == "" {
		return lockOwner{}, fmt.Errorf("incomplete lock metadata")
	}
	return owner, nil
}
